src/protocol.rs

```diff
@@ -511,6 +511,21 @@ impl ProofContext for SharedContext {
 }
 
 impl SharedContext {
+    #[cfg(test)]
+    pub fn gen_shared_context(
```

Comment:
nit: I would consolidate this function with your `generate_shared_context` function from the tests. I'd say to keep it here, rename it `random`, and generate all of the fields right here (with an RNG as a parameter). Maybe add a comment to emphasize that it shouldn't be used outside of tests.